plutus-use-cases/src/Plutus/Contracts/SealedBidAuction.hs

```diff
@@ -0,0 +1,304 @@
+   salt to the secret bids or using some other more sophisticated mechanism to
+   avoid this attack. In other words, please don't blidnly deploy this code
+   without understanding the possible attack scenarios.
+ -}
+
+data BidArgs = BidArgs{ secretBid :: SecretArgument Integer }
+          deriving stock (Haskell.Show, Generic)
+          deriving anyclass (ToJSON, FromJSON)
+
+data RevealArgs = RevealArgs{ publicBid :: Integer }
+          deriving stock (Haskell.Show, Generic)
+          deriving anyclass (ToJSON, FromJSON)
+
+type BidderSchema = Endpoint "bid" BidArgs
+                    .\/ Endpoint "reveal" RevealArgs
+                    .\/ Endpoint "payout" ()
+type SellerSchema = Endpoint "payout" ()
+
+-- | Definition of an auction
+data AuctionParams
+    = AuctionParams
+        { apOwner      :: PubKeyHash -- ^ Current owner of the asset. This is where the proceeds of the auction will be sent.
+        , apAsset      :: Value -- ^ The asset itself. This value is going to be locked by the auction script output.
+        , apEndTime    :: POSIXTime -- ^ When the time window for bidding ends.
+        , apPayoutTime :: POSIXTime -- ^ When the time window for revealing your bid ends.
+        }
+        deriving stock (Haskell.Eq, Haskell.Show, Generic)
+        deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.makeLift ''AuctionParams
+
+
+data SealedBid =
+    SealedBid
+        { sealedBid       :: BuiltinByteString
+        , sealedBidBidder :: PubKeyHash
+        }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''SealedBid
+
+instance Eq SealedBid where
+  (SealedBid bid bidder) == (SealedBid bid' bidder') = bid == bid' && bidder == bidder'
+
+data RevealedBid =
+    RevealedBid
+        { revealedBid       :: Integer
+        , revealedBidBidder :: PubKeyHash
+        }
+    deriving stock (Haskell.Eq, Haskell.Show, Generic)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''RevealedBid
+
+-- | The states of the auction
+data AuctionState
+    = Ongoing [SealedBid] -- Bids can be submitted.
+    | AwaitingPayout RevealedBid [SealedBid] -- The bidding is finished and we are awaiting payout
+    | Finished
+    deriving stock (Generic, Haskell.Show, Haskell.Eq)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''AuctionState
+
+-- | Initial 'AuctionState'. In the beginning there are no bids.
+initialState :: AuctionState
+initialState = Ongoing []
+
+-- | Transition between auction states
+data AuctionInput
+    = PlaceBid SealedBid -- Register a sealed bid
+    | RevealBid RevealedBid -- Reveal a bid
+    | Payout
+    deriving stock (Generic, Haskell.Show)
+    deriving anyclass (ToJSON, FromJSON)
+
+PlutusTx.unstableMakeIsData ''AuctionInput
+
+type AuctionMachine = StateMachine AuctionState AuctionInput
+
+{-# INLINABLE showInteger #-}
+showInteger :: Integer -> BuiltinString
```

Comment:
I didn't pay attention to this before. This is quite evil :o I mean, it's probably correct, I just expect it will perform dreadfully.

Reply:
@michaelpj and @j-mueller I didn't measure it but as you say it probably does perform dreadfully. A better way to hash integers would be to just plunk the bytes of the integer into a `BuiltinByteString` directly in X-endian (doesn't matter) order and hash that. I don't understand the plutus cost model well enough to make this fast however. Any suggestion?

Reply:
What I would like is to use `consByteString` - but I suppose this assumes the integer you cons is no more than one word (and the rest of this comment is using that assumption). If there was a plutus builtin function `words :: BuiltinInteger -> [BuiltinInteger]` that would of course mean that I have everything I need to make this fast. However, now ill just hack this together myself and assume that modulo and divide on powers of two is fast in Plutus.

Reply:
Anyway, I addressed this in `ab4a321`.